R/modify.R

```diff
@@ -93,41 +93,18 @@ modify <- function(.x, .f, ...) {
 modify.default <- function(.x, .f, ...) {
   .f <- as_mapper(.f, ...)
 
-  for (i in seq_along(.x)) {
-    list_slice2(.x, i) <- .f(.x[[i]], ...)
+  if (vec_is_list(.x) || is.data.frame(.x)) {
+    out <- map(vec_proxy(.x), .f, ...)
+    vec_restore(out, .x)
```

Comment:
For the df case you need to recreate a data frame with the correct number of rows before calling `vec_restore()`.

Reply:
Can you suggest a test that illustrates why what I'm currently doing is incorrect, because this currently appears to modify data frames just fine.

Reply:
It's to support the case of a data frame with 0 columns but positive number of rows.

Reply:
``` r
x <- data_frame(.size = 5)
modify(x, identity)
#> data frame with 0 columns and 0 rows
```

Reply:
I'd probably have a separate data frame branch like this


```r
if (is.data.frame(.x)) {
  size <- vec_size(.x)
  out <- vec_proxy(.x)
  out <- map(out, .f, ...)
  out <- new_data_frame(out, n = size)
  vec_restore(out, .x)
}
```

Reply:
Hmm actually that wouldn't work because then people can't use `modify()` to summarise a data frame like this

```r
df <- data.frame(x = 1:5, y = 6:10)
modify(df, sum)
```

@lionel- I'm not sure we can do this reliably for 0 column data frames. I think the output `size` is going to have to be purely determined by the results of the `.f` calls, and not from the `size` of `.x`. In other words, I think we can only maintain a length invariant, not a size invariant.

Reply:
Maybe I should just remove the `is.data.frame()` test and it will fail naturally into the `is.list()` branch, which will preserve the existing behaviour.

Reply:
But what should happen with the zero-column case where the size can't be determined from outputs? Should we create a brand new zero-row data frame or should we restore the size? I lean towards the second but I don't know what makes most sense here.

Reply:
Ok, not naturally, but I could rearrange the branches to:

```R
  if (is.null(.x)) {
    NULL
  } else if (vec_is_list(.x)) {
    out <- map(vec_proxy(.x), .f, ...)
    vec_restore(out, .x)
  } else if (is.list(.x)) {
    .x[] <- map(.x, .f, ...)
    .x
  } else if (vec_is(.x)) {
    map_vec(.x, .f, ..., .ptype = .x)
  } else {
    cli::cli_abort("Don't know how to modify {.obj_type_friendly {.x}}.")
  }
```

Reply:
Behaviour with CRAN purrr:

``` r
df <- data.frame(x = 1:5)
df |> purrr::modify(sum)
#>    x
#> 1 15
#> 2 15
#> 3 15
#> 4 15
#> 5 15
```

<sup>Created on 2022-09-16 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
Oh CRAN purrr retains the size of the input when it is a data frame? Nice, then you could do this:

```r
if (is.data.frame(.x)) {
  size <- vec_size(.x)
  out <- vec_proxy(.x)
  out <- map(out, .f, ...)
  out <- vec_recycle_common(!!!out, .size = size)
  out <- new_data_frame(out, n = size)
  vec_restore(out, .x)
}
```

Reply:
I just pushed a concrete proposal — I think it's worth explicitly recycling here in order to get a better error message.

Reply:
That fixes all of the cases I brought up:

``` r
x <- data_frame(.size = 5)
modify(x, identity)
#> data frame with 0 columns and 5 rows

df <- data.frame(x = 1:5, y = 6:10)
modify(df, sum)
#>    x  y
#> 1 15 40
#> 2 15 40
#> 3 15 40
#> 4 15 40
#> 5 15 40

modify(
  data.frame(x = 1, y = 2), 
  \(x) if (x == 1) 1:2 else 1
)
#> Error in `modify()`:
#> ! Can't recycle `x` (size 2) to size 1.
```